File: 2558. Take Gifts From the Richest Pile.cpp
class Solution
{
public:
    long long pickGifts(vector<int> &gifts, int k)
    {
        priority_queue<int> pq;
        for (auto val : gifts)
            pq.push(val);

        while (k--)
        {
            pq.push(sqrt(pq.top()));
            pq.pop();
        }

        long ans = 0;
        while (!pq.empty())
        {
            ans += pq.top();
            pq.pop();
        }
        return ans;
    }
};

File: 2558. Take Gifts From the Richest Pile.py
class Solution:
    def pickGifts(self, gifts: List[int], k: int) -> int:
        Heap = [-gift for gift in gifts]
        heapify(Heap)
        for _ in range(k):
            top = -heappop(Heap)
            heappush(Heap, -int(math.sqrt(top)))
        return -sum(Heap)